src/compiler/checker.ts

```diff
@@ -24853,7 +24853,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const templateType = getTemplateTypeFromMappedType(target);
         const inference = createInferenceInfo(typeParameter);
         inferTypes([inference], sourceType, templateType);
-        return getTypeFromInference(inference) || unknownType;
+        return getTypeFromInference(inference) || getBaseConstraintOfType(typeParameter) || unknownType;

```

Comment:
`getBaseConstraintOfType` recursively walks all constraints - so if `T extends U` and `U extends V` and `V extends number`, you get `number` out. Which means you skip inferences to those intermediate results. I think you'd want to use `getConstraintOfType`, right? Because, in this case, if you have `T[K] extends U` and `U extends number`, `U` is a better (and more specific) answer than `number`.

Reply:
> I think you'd want to use getConstraintOfType, right?

Sort of. It's true it won't immediately walk all constraints but at the same time, it's not inference-aware anyway. If I call it then I end up with `T[keyof T]` and when relating the inferred type to the instantiated constraint (from within `getInferredType`) we eventually simplify this to `U` and to its base constraint. So to some extent, this leads to the same thing and the same problems. To incorporate inferences of other type parameters I'd have to utilize `context.nonFixingMapper`. 

It's not immediately obvious how to get access to this from within this function (`inferReverseMappedType`) but that certainly can be done. A bigger problem though is that `inferReverseMappedType` gets called at different times for object and array/tuple types - when dealing with the latter it's called eagerly~. So it gets called right from within `inferFromTypes`. This introduces differences in data availability between the object and the array/tuple case.

For instance, let's take a look at this:
```ts
declare function foo<T extends U[], U extends string | number | boolean>(
  b: {
    [K in keyof T]: (arg: T[K]) => void;
  },
  a: U,
): T;

declare const data: number | string;

const result = foo([(arg) => {}, () => {}], data);
```

Since `inferReverseMappedType` gets called eagerly here to create the reverse mapped tuple type it's not possible to observe the `U`'s inferred type. That is different from the equivalent object variant, like:
```ts
declare function foo<
  T extends Record<PropertyKey, U>,
  U extends string | number | boolean,
>(
  b: {
    [K in keyof T]: (arg: T[K]) => void;
  },
  a: U,
): T;

declare const data: number | string;

const result = foo(
  {
    a: (arg) => {},
    b: () => {},
  },
  data,
);
```

This isn't exactly a new problem. I've encountered this already in the past at least once or twice.

I'd really like to solve this. It feels like a separate issue though - but then: what would be the acceptable state of this PR to get this one in? 
1. does anything have to be done to move this one forward?
2. would making `inferReverseMappedType` aware of the inference context (to get access to its `.nonFixingMapper`) be enough?
3. should the object/array/tuple be unified for the whole thing to work in the same predictable way?